<div style="
    background:linear-gradient(90deg, #1B4F72, #2874A6);
    padding:25px;
    border-radius:15px;
    text-align:center;
">

<h1 style="
    color:white;
    font-size:34px;
    margin:0;
    text-align:center !important;
">
    Customer Account Engagement Analysis
</h1>

<p style="
    color:#D6EAF8;
    font-size:16px;
    margin:5px 0 0;
    text-align:center !important;
">
    Exploring Customer Journeys, Activation and Transaction Engagement
</p>

</div>

## Project Overview

This project analyses anonymised customer account-level data from a UK prepaid card service. The service is designed for children aged 6–18, where an adult account holder applies for and manages prepaid cards on behalf of children. The adult account holder may be a biological parent, grandparent, elder sibling, or another eligible adult.

The service permits a maximum of four active cards at a time. However, an account holder may have more than four child accounts historically if previously issued cards were cancelled.

## Project Objective

The primary objective of this project is to understand customer account engagement and identify patterns in the customer journey. The analysis focuses on:

- Account creation patterns.
- The time taken between reliable customer lifecycle stages.
- Differences in engagement between product types.
- The relationship between account status and transaction activity.
- Identifying accounts with limited or no engagement.

## Dataset Description

The dataset contains anonymised account-level records and includes information such as:

- Account creation date.
- Card activation date.
- First top-up date.
- First transaction date.
- Last transaction date.
- Account status. (Verified/Unverified)
- Product type. (FDD: Full Product/ SDD: Simplified Product)
- Total transaction count.
- Total transaction value.
- Number of associated children.

The data will be cleaned and prepared before conducting exploratory data analysis using Python.
NB: During data quality assessment, `card_activation_date` was identified as unreliable and excluded from the final analysis dataset.

## Tools and Methodology

This project is completed entirely using Python in Jupyter Notebook. The analysis will use Pandas and NumPy for data preparation and analysis, along with Matplotlib, Seaborn, and Plotly for visualisation.

The project follows these stages:

1. Data loading and initial inspection.
2. Data cleaning and preprocessing.
3. Exploratory data analysis.
4. Visualisation and statistical analysis.
5. Insight generation and recommendations.

The findings will help identify key patterns in customer account engagement and provide data-driven insights into the customer lifecycle.

*****************************************

<div style="
    background-color:#EAF2F8;
    padding:20px;
    border-radius:12px;
    text-align:center;
    border-left:6px solid #2E86C1;
">

<h1 style="color:#1B4F72; margin-bottom:10px;">
    Phase 1: Data Loading & Initial Inspection
</h1>
</div>

#### Step 1: 🔹Import required libraries

In [125]:
import pandas as pd
import numpy as np

#### Step 2: 🔹Load the dataset

In [126]:
df = pd.read_csv("combined_account_lifecycle.csv")

#### Step 3: 🔹Confirm dataset shape

In [127]:
df.shape

(22870, 11)

#### Step 4: 🔹Inspect column names

In [128]:
df.columns

Index(['account_id', 'account_created_date', 'card_activation_date',
       'first_topup_date', 'first_transaction_date', 'last_transaction_date',
       'account_status', 'number_of_children', 'product_type',
       'total_transaction_count', 'total_transaction_value'],
      dtype='object')

#### Step 5: 🔹Inspect sample records

In [129]:
df.head()

,account_id,account_created_date,card_activation_date,first_topup_date,first_transaction_date,last_transaction_date,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value
0,299166,2020-12-30T23:52:25.727Z,NaN,NaN,NaN,NaN,UNVERIFIED,0,FDD,0,0.00
1,299164,2020-12-30T23:35:27.289Z,2024-01-12T09:13:51.005Z,2021-01-08T08:44:55.834Z,2021-02-12T00:56:13.000Z,2024-04-14T23:03:03.000Z,VERIFIED,1,SDD,144,875.03
2,299163,2020-12-30T22:19:57.367Z,NaN,2021-01-06T11:05:04.408Z,2021-01-13T03:37:07.000Z,2021-03-12T09:41:09.000Z,VERIFIED,2,FDD,5,27.76
3,299162,2020-12-30T21:35:28.452Z,NaN,NaN,NaN,NaN,VERIFIED,0,SDD,0,0.00
4,299161,2020-12-30T21:22:05.023Z,NaN,2020-12-30T21:22:10.732Z,NaN,NaN,VERIFIED,0,SDD,0,0.00


In [130]:
df.tail()

,account_id,account_created_date,card_activation_date,first_topup_date,first_transaction_date,last_transaction_date,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value
22865,324977,2023-01-01T18:18:46.222Z,NaN,2023-01-01T18:20:25.581Z,NaN,NaN,VERIFIED,0,FDD,0,0.0
22866,324973,2023-01-01T13:37:57.826Z,NaN,2023-01-01T13:38:29.862Z,NaN,NaN,VERIFIED,0,FDD,0,0.0
22867,324972,2023-01-01T13:00:38.422Z,NaN,NaN,NaN,NaN,VERIFIED,0,NaN,0,0.0
22868,324971,2023-01-01T12:15:41.534Z,NaN,2023-01-01T12:17:56.390Z,NaN,NaN,VERIFIED,0,FDD,0,0.0
22869,324970,2023-01-01T01:30:53.051Z,NaN,2023-01-01T01:31:53.301Z,NaN,NaN,VERIFIED,0,FDD,0,0.0


#### Step 6: 🔹Inspect data types

In [131]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22870 entries, 0 to 22869
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   account_id               22870 non-null  int64  
 1   account_created_date     22870 non-null  object 
 2   card_activation_date     7072 non-null   object 
 3   first_topup_date         14579 non-null  object 
 4   first_transaction_date   6501 non-null   object 
 5   last_transaction_date    6501 non-null   object 
 6   account_status           22870 non-null  object 
 7   number_of_children       22870 non-null  int64  
 8   product_type             19812 non-null  object 
 9   total_transaction_count  22870 non-null  int64  
 10  total_transaction_value  22870 non-null  float64
dtypes: float64(1), int64(3), object(7)
memory usage: 1.9+ MB


#### Step 7: 🔹Generate summary statistics

In [132]:
df.describe()

,account_id,number_of_children,total_transaction_count,total_transaction_value
count,22870.00000,22870.000000,22870.000000,22870.000000
mean,303640.48251,0.479099,39.728334,275.064646
std,18192.97979,0.782798,134.444130,1007.735453
min,274299.00000,0.000000,0.000000,0.000000
25%,287302.00000,0.000000,0.000000,0.000000
50%,302443.50000,0.000000,0.000000,0.000000
75%,319833.75000,1.000000,6.000000,40.000000
max,336766.00000,8.000000,2735.000000,31150.330000


#### Step 8: 🔹Inspect categorical values

In [133]:
df['account_status'].unique()

array(['UNVERIFIED', 'VERIFIED'], dtype=object)

In [134]:
df['account_status'].value_counts()

account_status
VERIFIED      18118
UNVERIFIED     4752
Name: count, dtype: int64

In [135]:
df['product_type'].unique()

array(['FDD', 'SDD', nan], dtype=object)

In [136]:
df['product_type'].value_counts()

product_type
FDD    16229
SDD     3583
Name: count, dtype: int64

*********************************************

<div style="
    background-color:#EAF2F8;
    padding:20px;
    border-radius:12px;
    text-align:center;
    border-left:6px solid #2E86C1;
">

<h1 style="color:#1B4F72; margin-bottom:10px;">
    Phase 2: Data Quality Assessment
</h1>
</div>

#### Step 9: 🔍Check Missing Values
Identify the number of missing values in each column to understand the completeness of the dataset.

In [137]:
df.isnull().sum()

account_id                     0
account_created_date           0
card_activation_date       15798
first_topup_date            8291
first_transaction_date     16369
last_transaction_date      16369
account_status                 0
number_of_children             0
product_type                3058
total_transaction_count        0
total_transaction_value        0
dtype: int64

##### 📝Observation

The dataset contains missing values primarily in customer journey and product-related columns. Missing date values may represent accounts that did not progress to a particular stage of the customer journey.

During data validation, the `card_activation_date` column was found to be unreliable because activation information is not directly available in the source database. In some records, the reported activation date occurs after the first top-up date. Therefore, `card_activation_date` will be excluded from the analysis, while `first_topup_date` will be used as the earliest reliable post-creation engagement milestone.

#### Step 10: 🔍Check Duplicate Records
Identify duplicate rows to ensure that duplicate account records are not present in the dataset.

In [138]:
df.duplicated().sum()

np.int64(0)

##### 📝Observation

No completely duplicate records were found in the dataset. Each row contains a unique combination of values across all columns.

#### Step 11: 🔍Check Duplicate Account IDs
Verify whether each customer account appears only once in the dataset.

In [139]:
df['account_id'].duplicated().sum()

np.int64(0)

##### 📝Observation

No duplicate `account_id` values were found. Each customer account appears only once in the dataset, confirming that the dataset maintains one record per account.

#### Step 12: 🔍Check Invalid Numeric Values
Check for negative values in numeric columns where negative values are not logically expected.

In [140]:
numeric_columns = [
    'number_of_children',
    'total_transaction_count',
    'total_transaction_value'
]

for col in numeric_columns:
    print(f"{col}: {(df[col] < 0).sum()} negative values")

number_of_children: 0 negative values
total_transaction_count: 0 negative values
total_transaction_value: 0 negative values


##### 📝Observation

No negative values were found in `number_of_children`, `total_transaction_count`, or `total_transaction_value`. The numeric values are within the expected non-negative range.

#### Step 13: 🔍 Check Logical Date Consistency
Verify whether customer journey dates follow a logical chronological order.

In [141]:
date_columns = [
    'account_created_date',
    'first_topup_date',
    'first_transaction_date',
    'last_transaction_date'
]

for col in date_columns:
    df[col] = pd.to_datetime(df[col], errors='coerce')

print(df[date_columns].dtypes)

account_created_date      datetime64[ns, UTC]
first_topup_date          datetime64[ns, UTC]
first_transaction_date    datetime64[ns, UTC]
last_transaction_date     datetime64[ns, UTC]
dtype: object


In [142]:
invalid_topup_dates = (
    df['first_topup_date'].notna() &
    (df['first_topup_date'] < df['account_created_date'])
).sum()

invalid_first_transaction_dates = (
    df['first_transaction_date'].notna() &
    (df['first_transaction_date'] < df['account_created_date'])
).sum()

invalid_last_transaction_dates = (
    df['last_transaction_date'].notna() &
    (df['last_transaction_date'] < df['first_transaction_date'])
).sum()

print("First top-up before account creation:", invalid_topup_dates)
print("First transaction before account creation:", invalid_first_transaction_dates)
print("Last transaction before first transaction:", invalid_last_transaction_dates)

First top-up before account creation: 0
First transaction before account creation: 0
Last transaction before first transaction: 0


In [143]:
invalid_activation_dates = (
    df['card_activation_date'].notna() &
    df['first_topup_date'].notna() &
    (df['first_topup_date'] < df['card_activation_date'])
).sum()

print("First top-up before activation:", invalid_activation_dates)

First top-up before activation: 140


##### 📝Observation

The reliable customer journey dates follow the expected chronological order. No records were found where the first top-up occurred before account creation, the first transaction occurred before account creation, or the last transaction occurred before the first transaction.

However, the `card_activation_date` column was found to be unreliable. A total of 140 records contain an activation date earlier than the first top-up date. Since activation information is not directly available in the source database, this column will be excluded from further analysis. The `first_topup_date` will instead be used as the earliest reliable post-creation engagement milestone.

#### Step 14: 🔍 Check Date Ranges
Inspect the minimum and maximum dates for the reliable customer journey fields.

In [144]:
date_columns = [
    'account_created_date',
    'first_topup_date',
    'first_transaction_date',
    'last_transaction_date'
]

df[date_columns].agg(['min', 'max']).T

,min,max
account_created_date,2020-01-01 13:43:22.484000+00:00,2026-08-30 18:20:34.877000+00:00
first_topup_date,2020-01-01 13:43:28.409000+00:00,2026-09-12 16:26:00.064000+00:00
first_transaction_date,2020-01-13 00:04:09+00:00,2026-09-14 00:41:19+00:00
last_transaction_date,2020-01-20 05:59:46+00:00,2026-09-14 09:40:11+00:00


##### 📝Observation

The reliable customer journey dates range from January 2020 to September 2026. No obviously invalid or extreme dates were identified during the date-range assessment. The latest transaction date extends beyond the latest account creation date, which is expected because existing accounts may continue to engage with the service after new accounts are created.

#### Step 15: 🔍Check Transaction Data Consistency
Compare transaction counts and values with transaction dates to identify inconsistencies in the dataset.

In [145]:
positive_count_missing_date = (
    (df['total_transaction_count'] > 0) &
    (df['first_transaction_date'].isna())
).sum()

zero_count_existing_date = (
    (df['total_transaction_count'] == 0) &
    (df['first_transaction_date'].notna())
).sum()

positive_value_zero_count = (
    (df['total_transaction_value'] > 0) &
    (df['total_transaction_count'] == 0)
).sum()

print("Positive transaction count but missing first transaction date:",
      positive_count_missing_date)

print("Zero transaction count but existing first transaction date:",
      zero_count_existing_date)

print("Positive transaction value but zero transaction count:",
      positive_value_zero_count)

Positive transaction count but missing first transaction date: 0
Zero transaction count but existing first transaction date: 0
Positive transaction value but zero transaction count: 0


##### 📝Observation

The transaction-related fields were checked for logical inconsistencies.

- All accounts with a positive transaction count had a recorded first transaction date.
- No accounts with zero transactions had a first transaction date.
- No accounts had a positive transaction value with zero transactions.

All three checks returned zero inconsistencies, indicating that the transaction count, transaction dates, and transaction values are internally consistent.

#### Step 16: 🔍 Check Categorical Consistency
Identify unexpected categories, inconsistent spelling, or different capitalisation.

In [146]:
categorical_columns = ['account_status', 'product_type']

for col in categorical_columns:
    print(f"\n{col}:")
    print(df[col].value_counts(dropna=False))


account_status:
account_status
VERIFIED      18118
UNVERIFIED     4752
Name: count, dtype: int64

product_type:
product_type
FDD    16229
SDD     3583
NaN     3058
Name: count, dtype: int64


##### 📝Observation

The categorical columns were checked for unexpected categories, inconsistent spelling, and capitalization differences.

- `account_status` contains two consistent values: `VERIFIED` and `UNVERIFIED`.
- `product_type` contains two consistent values: `FDD` and `SDD`.
- No spelling or capitalization inconsistencies were identified.
- No missing values were found in these categorical columns.

<div style="
    background-color:#EAF2F8;
    padding:20px;
    border-radius:12px;
    text-align:center;
    border-left:6px solid #2E86C1;
">

<h1 style="color:#1B4F72; margin-bottom:10px;">
    Phase 3: Data Cleaning and Preparation
</h1>
</div>

#### Step 17: 🏷️Review the account population

In [147]:
population_summary = (
    df.groupby(['account_status', 'number_of_children'])
      .size()
      .reset_index(name='account_count')
)

population_summary

,account_status,number_of_children,account_count
0,UNVERIFIED,0,4571
1,UNVERIFIED,1,131
2,UNVERIFIED,2,38
3,UNVERIFIED,3,12
4,VERIFIED,0,10627
5,VERIFIED,1,4964
6,VERIFIED,2,1958
7,VERIFIED,3,459
8,VERIFIED,4,102
9,VERIFIED,5,4


## Business Context Clarification

During the data quality investigation, the meaning of certain fields was clarified with internal product knowledge.

The `account_status` field should not be used alone to identify invalid or incomplete customer accounts. `UNVERIFIED` accounts with one or more children can represent genuine parent accounts whose email has been verified but whose phone number verification is pending. Phone verification is completed when the parent installs the application and logs in to activate the card(s). Therefore, `account_status` will be treated as a descriptive customer-account attribute rather than an exclusion criterion.

Similarly, `number_of_children` represents the number of child accounts associated with a parent and does not necessarily represent the number of currently active cards. This particular prepaid card permits a maximum of four active cards at one time, but previously cancelled child cards may allow a parent to add additional children later. Therefore, parents with more than four historical child accounts are valid accounts and should not be excluded.

In this project, the term "parent" refers to the adult account holder who applies for and manages the account. This person may be a biological parent, grandparent, elder sibling, or another adult aged 18 or above.

## Defining the Analysis Population

For this analysis, the focus is on parent accounts associated with at least one child account. Records with `number_of_children = 0` will be excluded from the analysis population because they are outside the scope of this customer-card engagement analysis. Based on internal business validation, these records primarily represent incomplete applications or gifting accounts.

**The original dataset was preserved as `df`. A separate `analysis_df` will be created for subsequent analysis.**

The `card_activation_date` column will also be excluded from `analysis_df` because data validation identified inconsistencies between activation dates and first top-up dates. The first top-up date will be retained as the earliest reliable post-creation engagement milestone.

************************************************************

#### Step 18: 📄Create and Validate analysis_df

In [148]:
analysis_df = df[
    df['number_of_children'] > 0
].drop(columns=['card_activation_date']).copy()

In [149]:
analysis_df.shape

(7672, 10)

In [150]:
analysis_df.columns

Index(['account_id', 'account_created_date', 'first_topup_date',
       'first_transaction_date', 'last_transaction_date', 'account_status',
       'number_of_children', 'product_type', 'total_transaction_count',
       'total_transaction_value'],
      dtype='object')

In [151]:
analysis_df.duplicated().sum()

np.int64(0)

In [152]:
analysis_df['account_id'].duplicated().sum()

np.int64(0)

##### 📝Observation

After excluding accounts with `number_of_children = 0` and removing the unreliable `card_activation_date` column, the resulting `analysis_df` contains 7,672 parent accounts and 10 columns.

The dataset contains no duplicate records or duplicate account IDs.

Missing values remain in the transaction milestone date columns. These are considered meaningful because they may indicate accounts that have not yet completed a top-up or made a transaction. Therefore, these values were retained for further engagement analysis rather than imputed or removed.

**************************************************

#### Step 19: 🩹Handle Missing Values

In [153]:
# Inspect the current missing-value counts
analysis_df.isnull().sum()

account_id                    0
account_created_date          0
first_topup_date            991
first_transaction_date     1171
last_transaction_date      1171
account_status                0
number_of_children            0
product_type                  0
total_transaction_count       0
total_transaction_value       0
dtype: int64

In [154]:
# calculate missing-value percentages:
missing_summary = pd.DataFrame({
    'Missing Count': analysis_df.isnull().sum(),
    'Missing Percentage': (
        analysis_df.isnull().sum() / len(analysis_df) * 100
    ).round(2)
})

missing_summary

,Missing Count,Missing Percentage
account_id,0,0.00
account_created_date,0,0.00
first_topup_date,991,12.92
first_transaction_date,1171,15.26
last_transaction_date,1171,15.26
account_status,0,0.00
number_of_children,0,0.00
product_type,0,0.00
total_transaction_count,0,0.00
total_transaction_value,0,0.00


##### 📝Cleaning decision
Missing milestone dates are retained because they represent incomplete or absent customer activity rather than data-entry errors.

*******************************************

#### Step 20: 📊 Check Outliers

In [155]:
numerical_columns = [
    'number_of_children',
    'total_transaction_count',
    'total_transaction_value'
]

In [156]:
# Get a statistical overview
analysis_df[numerical_columns].describe()

,number_of_children,total_transaction_count,total_transaction_value
count,7672.000000,7672.000000,7672.000000
mean,1.428180,118.428962,819.959391
std,0.686426,211.103566,1606.449153
min,1.000000,0.000000,0.000000
25%,1.000000,6.000000,39.240000
50%,1.000000,37.000000,253.145000
75%,2.000000,135.000000,921.665000
max,8.000000,2735.000000,31150.330000


In [157]:
# Detect outliers using the IQR method
outlier_summary = []

for column in numerical_columns:
    Q1 = analysis_df[column].quantile(0.25)
    Q3 = analysis_df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outlier_count = (
        (analysis_df[column] < lower_bound) |
        (analysis_df[column] > upper_bound)
    ).sum()

    outlier_summary.append({
        'Column': column,
        'Q1': Q1,
        'Q3': Q3,
        'IQR': IQR,
        'Lower Bound': lower_bound,
        'Upper Bound': upper_bound,
        'Outlier Count': outlier_count
    })

outlier_summary_df = pd.DataFrame(outlier_summary)

outlier_summary_df

,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,number_of_children,1.00,2.000,1.000,-0.5000,3.5000,110
1,total_transaction_count,6.00,135.000,129.000,-187.5000,328.5000,767
2,total_transaction_value,39.24,921.665,882.425,-1284.3975,2245.3025,748


In [158]:
# Inspect accounts with high transaction counts

transaction_count_outliers = analysis_df[
    analysis_df['total_transaction_count'] > 328.5
].sort_values(
    'total_transaction_count',
    ascending=False
)

transaction_count_outliers[[
    'account_id',
    'account_status',
    'number_of_children',
    'product_type',
    'total_transaction_count',
    'total_transaction_value'
]].head(5)

,account_id,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value
10332,324741,VERIFIED,1,FDD,2735,15418.44
9391,276226,VERIFIED,2,FDD,2647,22181.54
4562,287102,VERIFIED,1,FDD,2414,23310.47
2675,292326,VERIFIED,1,FDD,2225,11107.05
5657,284355,VERIFIED,2,FDD,2168,18549.76


In [159]:
# Inspect accounts with high transaction values

transaction_value_outliers = analysis_df[
    analysis_df['total_transaction_value'] > 2245.3025
].sort_values(
    'total_transaction_value',
    ascending=False
)

transaction_value_outliers[[
    'account_id',
    'account_status',
    'number_of_children',
    'product_type',
    'total_transaction_count',
    'total_transaction_value'
]].head(5)

,account_id,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value
12220,319605,VERIFIED,1,FDD,1491,31150.33
5498,284804,VERIFIED,1,FDD,1285,25328.71
12219,319607,VERIFIED,1,FDD,539,24464.81
4562,287102,VERIFIED,1,FDD,2414,23310.47
9391,276226,VERIFIED,2,FDD,2647,22181.54


##### 📊 Outlier Treatment

Potential outliers were identified in the numerical columns using the Interquartile Range (IQR) method.

The analysis identified potential outliers in:

- `number_of_children`
- `total_transaction_count`
- `total_transaction_value`

These outliers were reviewed and determined to represent genuine customer accounts and valid customer behaviour rather than data-quality errors. For example, some customers have a higher number of children or significantly higher transaction activity than the typical account.

Therefore, no records were removed or modified based solely on the IQR outlier detection results.

**Decision:** All identified outliers have been retained to preserve the accuracy and completeness of the customer engagement analysis.

***************************************

<div style="
    background-color:#EAF2F8;
    padding:20px;
    border-radius:12px;
    text-align:center;
    border-left:6px solid #2E86C1;
">

<h1 style="color:#1B4F72; margin-bottom:10px;">
    Phase 4: Feature Engineering
</h1>
</div>

#### Step 21: ➕ Create Derived Columns

##### Derived Columns and Their Purpose

The following derived columns were created to support customer journey, engagement, transaction, and data-quality analysis.

| Derived Column | Purpose |
|---|---|
| `days_to_first_topup` | Measures the number of days between account creation and the first recorded top-up. |
| `has_first_topup` | Indicates whether a first top-up has been recorded for the account. |
| `has_first_transaction` | Indicates whether a first transaction has been recorded for the account. |
| `days_to_first_transaction` | Measures the number of days between account creation and the first recorded transaction. |
| `transaction_value_per_transaction` | Calculates the average recorded transaction value per transaction. |
| `transaction_activity_span_days` | Measures the number of days between the first and last recorded transactions. |
| `days_from_topup_to_first_transaction` | Measures the number of days between the first recorded top-up and first recorded transaction. |
| `days_from_topup_to_first_transaction_clean` | Provides a cleaned version of the top-up-to-first-transaction duration, excluding negative values from analytical calculations. |
| `invalid_topup_transaction_sequence` | Flags records where the recorded first transaction occurred before the recorded first top-up. |
| `account_creation_year` | Extracts the year from the account creation date. |
| `account_creation_month` | Extracts the month from the account creation date. |
| `account_creation_quarter` | Extracts the quarter from the account creation date. |
| `customer_lifecycle_stage` | Categorises accounts based on their progression from account creation to recorded transaction activity. |
| `transaction_activity_segment` | Groups accounts based on the number of recorded transactions. |
| `transaction_value_segment` | Groups accounts based on their total recorded transaction value. |

In [160]:
# days_to_first_topup

analysis_df['days_to_first_topup'] = (
    analysis_df['first_topup_date'] -
    analysis_df['account_created_date']
).dt.days

In [161]:
# has_first_topup

analysis_df['has_first_topup'] = (
    analysis_df['first_topup_date'].notna()
)

In [162]:
analysis_df['has_first_topup'].value_counts()

has_first_topup
True     6681
False     991
Name: count, dtype: int64

In [163]:
# has_first_transaction

analysis_df['has_first_transaction'] = (
    analysis_df['first_transaction_date'].notna()
)

In [164]:
analysis_df['has_first_transaction'].value_counts()

has_first_transaction
True     6501
False    1171
Name: count, dtype: int64

In [165]:
# days_to_first_transaction

analysis_df['days_to_first_transaction'] = (
    analysis_df['first_transaction_date'] -
    analysis_df['account_created_date']
).dt.days

In [166]:
analysis_df['days_to_first_transaction'].describe()

count    6501.000000
mean       32.730657
std        68.533353
min         4.000000
25%        10.000000
50%        15.000000
75%        29.000000
max      1338.000000
Name: days_to_first_transaction, dtype: float64

In [167]:
# transaction_value_per_transaction

analysis_df['transaction_value_per_transaction'] = (
    analysis_df['total_transaction_value'] /
    analysis_df['total_transaction_count']
)

In [168]:
analysis_df['transaction_value_per_transaction'].describe()

count    6501.000000
mean        8.455851
std         8.185008
min         0.824444
25%         4.346630
50%         6.424108
75%         9.790000
max       132.954545
Name: transaction_value_per_transaction, dtype: float64

In [169]:
# transaction_activity_span_days

analysis_df['transaction_activity_span_days'] = (
    analysis_df['last_transaction_date'] -
    analysis_df['first_transaction_date']
).dt.days

In [170]:
analysis_df['transaction_activity_span_days'].describe()

count    6501.000000
mean      518.937087
std       510.260218
min         0.000000
25%       103.000000
50%       363.000000
75%       779.000000
max      2397.000000
Name: transaction_activity_span_days, dtype: float64

In [171]:
# days_from_topup_to_first_transaction
analysis_df['days_from_topup_to_first_transaction'] = (
    analysis_df['first_transaction_date'] -
    analysis_df['first_topup_date']
).dt.days

In [172]:
analysis_df[
    [
        'first_topup_date',
        'first_transaction_date',
        'days_from_topup_to_first_transaction'
    ]
].head(10)

,first_topup_date,first_transaction_date,days_from_topup_to_first_transaction
1,2021-01-08 08:44:55.834000+00:00,2021-02-12 00:56:13+00:00,34.0
2,2021-01-06 11:05:04.408000+00:00,2021-01-13 03:37:07+00:00,6.0
6,NaT,NaT,NaN
8,2021-01-06 14:35:46.800000+00:00,2021-01-25 00:20:23+00:00,18.0
9,2021-01-06 16:45:02.887000+00:00,2021-01-07 09:41:11+00:00,0.0
15,2021-01-18 02:22:31.427000+00:00,2021-01-13 00:56:10+00:00,-6.0
17,2021-01-08 11:15:47.170000+00:00,2021-01-12 00:56:19+00:00,3.0
23,2021-01-07 13:19:54.136000+00:00,2021-01-21 03:37:41+00:00,13.0
25,2021-01-06 14:57:04.730000+00:00,2021-01-07 00:56:11+00:00,0.0
26,2021-01-10 20:54:54.622000+00:00,2021-03-17 02:41:07+00:00,65.0


In [173]:
(analysis_df['days_from_topup_to_first_transaction'] < 0).sum()

np.int64(134)

##### 📝Observation: Negative Duration Between Top-up and First Transaction
134 records contained negative durations between the recorded first top-up and first transaction dates. This may mainly indicate data inconsistencies, although a small number of legitimate exceptions, such as an initial gifter load followed by child spending before the parent’s recorded top-up, cannot be ruled out. Since the exact cause cannot be confirmed for all records, the affected rows were retained and flagged for review.

In [174]:
# Creating a new column 'days_from_topup_to_first_transaction_clean'
analysis_df['days_from_topup_to_first_transaction_clean'] = (
    analysis_df['days_from_topup_to_first_transaction']
    .where(
        analysis_df['days_from_topup_to_first_transaction'] >= 0
    )
)

In [175]:
# Create a new column 'invalid_topup_transaction_sequence' to Flag the problematic records
analysis_df['invalid_topup_transaction_sequence'] = (
    analysis_df['days_from_topup_to_first_transaction'] < 0
)

In [176]:
# Column: account_creation_year
analysis_df['account_creation_year'] = (
    analysis_df['account_created_date'].dt.year
)

In [177]:
# Column: account_creation_month
analysis_df['account_creation_month'] = (
    analysis_df['account_created_date'].dt.month
)

In [178]:
# Column: account_creation_quarter
analysis_df['account_creation_quarter'] = (
    analysis_df['account_created_date'].dt.quarter
)

### Feature Engineering: Customer Segmentation

Three categorical features will be created to analyse customer progression and transaction behaviour.

| Feature | Category | Definition |
|---|---|---|
| Customer Lifecycle Stage | `Created_No_Topup` | Account created, but no first top-up recorded. |
|  | `Topup_No_Transaction` | First top-up recorded, but no first transaction recorded. |
|  | `First_Transaction` | One recorded transaction. |
|  | `Repeat_Transaction` | More than one recorded transaction. |
| Transaction Activity Segment | `No_Activity` | 0 recorded transactions. |
|  | `Low_Activity` | 1–5 recorded transactions. |
|  | `Medium_Activity` | 6–20 recorded transactions. |
|  | `High_Activity` | More than 20 recorded transactions. |
| Transaction Value Segment | `No_Value` | £0 recorded transaction value. |
|  | `Low_Value` | Greater than £0 and up to £50. |
|  | `Medium_Value` | Greater than £50 and up to £200. |
|  | `High_Value` | Greater than £200. |

> The segmentation thresholds are initial analytical categories and may be reviewed during EDA. All categories are based on recorded historical activity and do not indicate current account status.

In [179]:
def assign_lifecycle_stage(row):
    if not row['has_first_topup']:
        return 'Created_No_Topup'
    elif not row['has_first_transaction']:
        return 'Topup_No_Transaction'
    elif row['total_transaction_count'] == 1:
        return 'First_Transaction'
    else:
        return 'Repeat_Transaction'

analysis_df['customer_lifecycle_stage'] = analysis_df.apply(
    assign_lifecycle_stage,
    axis=1
)

In [180]:
def assign_activity_segment(count):
    if count == 0:
        return 'No_Activity'
    elif count <= 5:
        return 'Low_Activity'
    elif count <= 20:
        return 'Medium_Activity'
    else:
        return 'High_Activity'

analysis_df['transaction_activity_segment'] = (
    analysis_df['total_transaction_count']
    .apply(assign_activity_segment)
)

In [181]:
def assign_value_segment(value):
    if value == 0:
        return 'No_Value'
    elif value <= 50:
        return 'Low_Value'
    elif value <= 200:
        return 'Medium_Value'
    else:
        return 'High_Value'

analysis_df['transaction_value_segment'] = (
    analysis_df['total_transaction_value']
    .apply(assign_value_segment)
)

#### Step 22: ✅ Validate cleaned dataset

In [182]:
# 1. Check Dataset Shape
analysis_df.shape

(7672, 25)

In [183]:
# 2. Check Duplicate Records
analysis_df.duplicated().sum()

np.int64(0)

In [184]:
analysis_df['account_id'].duplicated().sum()

np.int64(0)

In [185]:
# 3. Check Missing Values
analysis_df.isnull().sum()

account_id                                       0
account_created_date                             0
first_topup_date                               991
first_transaction_date                        1171
last_transaction_date                         1171
account_status                                   0
number_of_children                               0
product_type                                     0
total_transaction_count                          0
total_transaction_value                          0
days_to_first_topup                            991
has_first_topup                                  0
has_first_transaction                            0
days_to_first_transaction                     1171
transaction_value_per_transaction             1171
transaction_activity_span_days                1171
days_from_topup_to_first_transaction          1201
days_from_topup_to_first_transaction_clean    1335
invalid_topup_transaction_sequence               0
account_creation_year          

In [186]:
# 4. Check Data Types
analysis_df.dtypes

account_id                                                  int64
account_created_date                          datetime64[ns, UTC]
first_topup_date                              datetime64[ns, UTC]
first_transaction_date                        datetime64[ns, UTC]
last_transaction_date                         datetime64[ns, UTC]
account_status                                             object
number_of_children                                          int64
product_type                                               object
total_transaction_count                                     int64
total_transaction_value                                   float64
days_to_first_topup                                       float64
has_first_topup                                              bool
has_first_transaction                                        bool
days_to_first_transaction                                 float64
transaction_value_per_transaction                         float64
transactio

In [187]:
# 5. Check Numerical Consistency
analysis_df[
    (analysis_df['total_transaction_count'] < 0) |
    (analysis_df['total_transaction_value'] < 0)
]

,account_id,account_created_date,first_topup_date,first_transaction_date,last_transaction_date,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value,...,transaction_activity_span_days,days_from_topup_to_first_transaction,days_from_topup_to_first_transaction_clean,invalid_topup_transaction_sequence,account_creation_year,account_creation_month,account_creation_quarter,customer_lifecycle_stage,transaction_activity_segment,transaction_value_segment


In [188]:
# Check whether transaction values are logically consistent:
analysis_df[
    (analysis_df['total_transaction_count'] == 0) &
    (analysis_df['total_transaction_value'] != 0)
]

,account_id,account_created_date,first_topup_date,first_transaction_date,last_transaction_date,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value,...,transaction_activity_span_days,days_from_topup_to_first_transaction,days_from_topup_to_first_transaction_clean,invalid_topup_transaction_sequence,account_creation_year,account_creation_month,account_creation_quarter,customer_lifecycle_stage,transaction_activity_segment,transaction_value_segment


In [189]:
# Check First Top-up Flag
analysis_df.groupby('has_first_topup')['first_topup_date'].count()

has_first_topup
False       0
True     6681
Name: first_topup_date, dtype: int64

In [190]:
# Check First Transaction Flag
analysis_df.groupby('has_first_transaction')['first_transaction_date'].count()

has_first_transaction
False       0
True     6501
Name: first_transaction_date, dtype: int64

In [196]:
# Check Negative Duration Values

negative_duration_check = analysis_df[
    (analysis_df['days_to_first_topup'] < 0) |
    (analysis_df['days_to_first_transaction'] < 0) |
    (analysis_df['transaction_activity_span_days'] < 0) |
    (analysis_df['days_from_topup_to_first_transaction'] < 0)
]

negative_duration_check.shape

(134, 25)

In [192]:
# Check Transaction Value per Transaction
analysis_df[
    analysis_df['transaction_value_per_transaction'] < 0
]

,account_id,account_created_date,first_topup_date,first_transaction_date,last_transaction_date,account_status,number_of_children,product_type,total_transaction_count,total_transaction_value,...,transaction_activity_span_days,days_from_topup_to_first_transaction,days_from_topup_to_first_transaction_clean,invalid_topup_transaction_sequence,account_creation_year,account_creation_month,account_creation_quarter,customer_lifecycle_stage,transaction_activity_segment,transaction_value_segment


In [197]:
validation_summary = pd.DataFrame({
    'Check': [
        'Dataset shape',
        'Duplicate rows',
        'Duplicate account IDs',
        'Missing values reviewed',
        'Negative numerical values',
        'Negative duration values'
    ],
    'Result': [
        analysis_df.shape,
        analysis_df.duplicated().sum(),
        analysis_df['account_id'].duplicated().sum(),
        'Expected missing milestone dates retained',
        (
            (analysis_df['total_transaction_count'] < 0).sum() +
            (analysis_df['total_transaction_value'] < 0).sum()
        ),
        (
         (analysis_df['days_to_first_topup'] < 0).sum() +
         (analysis_df['days_to_first_transaction'] < 0).sum() +
         (analysis_df['transaction_activity_span_days'] < 0).sum() +
         (analysis_df['days_from_topup_to_first_transaction'] < 0).sum()
        )
    ]
})

validation_summary

,Check,Result
0,Dataset shape,"(7672, 25)"
1,Duplicate rows,0
2,Duplicate account IDs,0
3,Missing values reviewed,Expected missing milestone dates retained
4,Negative numerical values,0
5,Negative duration values,134


### Validate Cleaned Dataset

The cleaned and feature-engineered dataset was validated to ensure data quality and consistency before proceeding to EDA.

#### Validation Checks

| Check | Result |
|---|---|
| Dataset shape | 7,672 rows × 25 columns |
| Duplicate rows | None found |
| Duplicate account IDs | None found |
| Negative numerical values | None found |
| Negative duration values | 134 records flagged for review |
| Missing values | Reviewed and retained where meaningful |
| Data types | Verified and appropriate for analysis |

#### Missing Value Observations

Missing values were retained where they represent unavailable customer journey milestones rather than data-entry errors.

- `first_topup_date`: 991 missing values
- `first_transaction_date`: 1,171 missing values
- `last_transaction_date`: 1,171 missing values
- Derived duration columns contain missing values when the required source dates are unavailable.
- `days_from_topup_to_first_transaction_clean` excludes negative durations from analytical calculations.

#### Feature Engineering Validation

The following derived features were successfully created:

- Customer journey duration metrics
- First top-up and first transaction indicators
- Transaction value per transaction
- Transaction activity span
- Account creation year, month, and quarter
- Customer lifecycle stage
- Transaction activity segment
- Transaction value segment
- Data-quality flag for inconsistent top-up and transaction sequences

The dataset is now cleaned, validated, and feature-engineered for the next phase: Exploratory Data Analysis (EDA) and Visualization.

***********************************************

#### Step 23: 💾 Export Cleaned Dataset

After completing the data cleaning, validation, and feature engineering steps, the final prepared dataset was exported as a CSV file.

The dataset contains 7,672 parent accounts and 25 columns, including the original cleaned fields and derived customer engagement features.

CSV format was selected to preserve timezone-aware datetime values and maintain compatibility with Python, Excel, Power BI, and other analytical tools.

The exported file will serve as the working dataset for the next phase: Exploratory Data Analysis (EDA) and Visualization.

In [195]:
# Export cleaned dataset as CSV

analysis_df.to_csv(
    "DA_Final_Project_Cleaned_Dataset.csv",
    index=False
)